### ISSUE TITLE
🫱🏻‍🫲🏽 Merge with MegaLinter ?

### ISSUE THREAD

**Issue by nvuillam** [super-linter/super-linter#4591]
Hi dear Super-Linter maintainers & contributors

I know this topic has already been discussed years ago, but I think the current situation deserves a new discussion :)

MegaLinter was originally a hard-fork of Super-Linter, after had been declined a huge PR that rewrote everything from bash to python, for reasons I totally understand for that time.

But some time later, here is the situation I think I observe:

- MegaLinter is still growing ( **5 millions** docker pulls , [1400 stars](https://github.com/oxsecurity/megalinter/stargazers/), [used by 2700+ repos on Github only](https://megalinter.io/latest/used-by-stats/), [almost 100 contributors](https://github.com/oxsecurity/megalinter/graphs/contributors) , used by OSS repos of Microsoft, AWS & Azure) , and has many active maintainers ( @Kurt-von-Laven , @bdovaz ,  @echoix & myself ) and great contributors like @DariuszPorowski & @ThomasSanson .

- Super-Linter seems stagnating, and the fact that it is not on github organization anymore won't help, and it seems original creators are less active

## Super-Linter stats

![image](https://github.com/super-linter/super-linter/assets/17500430/378cb5ac-bc63-4d81-8925-020d67d5b328)
![image](https://github.com/super-linter/super-linter/assets/17500430/6bc5103f-545f-477c-9023-6422e2a5004f)
![image](https://github.com/super-linter/super-linter/assets/17500430/b6a8e6f1-37d3-409e-b379-a2eefb5934e4)

## MegaLinter stats

![image](https://github.com/super-linter/super-linter/assets/17500430/4be4fc7d-e105-4de5-90e5-3354c7440554)
![image](https://github.com/super-linter/super-linter/assets/17500430/6a19d953-b20d-4a34-b44b-8b165688cb24)
![image](https://github.com/super-linter/super-linter/assets/17500430/0e0f84f8-bb08-4c4f-b589-7a1fb3abe294)

To summarize, we are maintaining similar products, that can handle the same configuration, and python makes that capabilities of MegaLinter are way ahead of Super-Linter capabilities. -> https://megalinter.io/beta/mega-linter-vs-super-linter/

Why not joining forces and at the same time save us some time by sharing the effort to provide open-source free linters aggregator to the developer community ?

MegaLinter is now backed by [OX Security](https://ox.security/), but I can personnally assure that it will always be open-source and free for everyone, we share common values with OX Security Founders, there is 0% risk that someday someone gets crazy and decides to make it closed-source or not free anymore.

So, let's work together on making MegaLinter even greater that it already is thanks to the original work on SuperLinter ? :)

Best regards

[Nicolas Vuillamy](https://www.linkedin.com/in/nicolas-vuillamy/) (please feel free to directly contact me if you'd like an informal discussion :))

**Comment by Hanse00** [comment]
I won't speak on behalf of anyone but myself, but in a that capacity I will say: All else being the same, not duplicating work sounds like a win.

I'm open to having the discussion if @zkoppert is.

**Comment by nvuillam** [comment]
@Hanse00 many thanks for this first feedback :)

I thought a little about how it could eventually go... it would generate work on both sides to organize such migration, but also save some work on SL side because ML already have many features that are requested in SL issues:

- Custom scripts https://github.com/super-linter/super-linter/issues/4525 -> [PRE_COMMANDS](https://megalinter.io/latest/config-precommands/) , [POST_COMMANDS](https://megalinter.io/latest/config-postcommands/)
- Upate dotnet version https://github.com/super-linter/super-linter/issues/4522 -> ML is using dotnet v6
- Probably more ^^

We also are working on [Multi-arch](https://github.com/super-linter/super-linter/issues/4498) for a while... but it's [not published yet](https://github.com/oxsecurity/megalinter/pull/2273), maybe with your help it would go faster ^^

On ML side, a few linters existing in SL are missing and should be added to ensure compatibility:
- clang-format
- google-java-format
- textlint

**Comment by nvuillam** [comment]
@zkoppert any thoughts ? 🙂

**Comment by zkoppert** [comment]
Yes, sorry for the delays. I've been away from my computer for vacation. I see the value in joining forces to maintain a similar tool. I think we could further discuss what a merge of the projects would look like: governance structure, time commitments, resources, etc.

As well, I'm curious what role ox security plays. Is it funding resources, time, or something else?

@ferrarimarco any thoughts?

**Comment by ferrarimarco** [comment]
Hi folks!

This is probably going to add questions, rather than answers. :)

I don't know what this implies on the project side, that's probably for GitHub folks to answer (@zkoppert probably).

On the technical side, this initiative looks, to me, a deprecation of super-linter in favor of mega-linter. Not having inspected the mega-linter code base and tooling, I don't know. if this is a step forward or backward, or both :)

**Comment by nvuillam** [comment]
@zkoppert > You're so right to have enjoyed your vacations without your computer ;)

About what it would like, everything is open to discussion, but as @ferrarimarco has guessed, I would imagine a deprecation of super-linter with a migration to MegaLinter when all the following conditions are met:

- ML contain the few missing delta features (clang-format, textlint, google-java-format)
- Aliasing in ML some SL config variables
- Ascending compatibility between super-linter & MegaLinter would have been tested, so SL users would not suffer from the migration and would just have a simple script to run to update their pipelines and scripts

About governance, I'm kind of a benevolent dictator with 3 maintainers (@Kurt-von-Laven ,@bdovaz, @echoix ) , but I (think I'm) easy to work with, you can discuss in private with them to have their opinion somewhere I don't read 🤣 
I would gladly be open to add SL active maintainers as ML maintainers (if my co-maintainers agree to that, because as said before I'm benevolent ^^ )

About time commitment and resources:
- We are an artistic open-source project so there is no commitement for anyone working on it: When you have time to work on it, you can work on it, but when you have a heavy charge at your job or prefer to go in vacations, nobody will ever blame you ^^
- The main resource is MegaLinter community: maintainers, contributors, and users who provide precious feedbacks !

About OX Security:
- I have an engagement with the 2 founders to maintain MegaLinter, especially on Security linters, because they are used in OX Security products
- MegaLinter is an OX Security branded product, but if one of the parties (me or OX) decide to stop the collaboration, MegaLinter ownership will come back to me (transfer from oxsecurity to nvuillam or any github org I'm owner of). Such thing is written in an official contract.
- I share common values about open-source with Neatsun & Lior, co-founders of OX Security, the collaboration with them is excellent since its beginning in early 2022, the trust is mutual and I have full independance about MegaLinter maintenance & roadmap
- OX Security has design & marketing departments, and they gladly provide their services when MegaLinter need it :)

Did I answer to your questions with theses informations ? :)

**Comment by Hanse00** [comment]
I guess my first high level question (Primarily directed at @zkoppert) would be: Are we empowered to make this decision, or is there someone else that should be a part of this conversation?

SuperLinter is a GitHub project, the license and code of conduct make that rather clear, even though the repo was moved out of the org. So are we in a position of authority to decide what happens to the project?

The code is MIT licensed of course, so I'm not suggesting there's a problem with MegaLinter being around. But I wonder if there are any company politics involved in us, on behalf of the project, making any kind of "official" statement around MegaLinter succeeding this project.

I'm asking because I get the impression GitHub might be a company that's fairly relaxed about this stuff, but I've also worked at places where the corporate legal council would want to get involved about something like this, so it would be good to feel out what our level of agency here is.

**Comment by zkoppert** [comment]
From a GitHub level, I did check and we are empowered to make this decision. 

Appreciate the other responses and input here as well. I'm continuing to think about this and trying to consider what is good for the community as well as consider how I personally feel about this. Feel free to continue to discuss here but I will probably take the next few weeks to keep thinking about my opinion on this.

**Comment by nvuillam** [comment]
@zkoppert there is no rush ^^

If I can add more pros to why it would be great to share the workload, these last few days we had to:
- Upgrade ktlint test cases because their new version finds more violations
- Downgrade kubescape because of a bug there
- Handle upgrade of tflint

And current urgent backlog contains elements like:
- Fix R lintr because its newest version seems to not work anymore
- create a "formatters only" flavor

Such tasks must probably also be done with SuperLinter... 

So in a bright future we could share those tasks instead of handling them in both repositories ^^

I see you are located in Oregon, it seems far but maybe not so far away from San Francisco, where I'll be for dreamforce all next week, if you even want to discuss in person :)

**Comment by zkoppert** [comment]
Highlights points that we would like to note as differences between the projects collected from discussion and comments above:
- Megalinter is licensed as AGPL3 while the super-linter is MIT
- Shared workload if we all were only managing 1 project instead of separately
- Megalinter is in Python and super-linter is in Shell/Bash
- Super linter receives additional features (Advanced security and Actions minutes) from GitHub, while OX-Security sponsors the Megalinter
- Megalinter is maintained by Nicolas as a BDFL model with other support maintainers while super linter is run as a community of equal maintainers
- Super linter is not financially sponsored, and Megalinter sponsorship is tied to Nicolas's github sponsors account
- The transition plan would be a deprecation of super-linter in favor of mega-linter, while porting forward some super linter features that megalinter might be missing
- Mega linter has higher repository and container usage stats compared to the megalinter when looking at recent periods of time



With that summary I would like to ask super-linter community members and maintainers to weigh in (especially those that haven't already) with their thoughts and opinions by commenting on this issue until October 15th, 2023. At that time we will consider all the community and maintainer thoughts and work to come to a decision to merge or stay separate.
cc/ Maintainers: @admiralawkbar @lindluni @zkoppert @nemchik @Hanse00 @GaboFDC @ferrarimarco 
cc/ Top Contributors: @assignUser @cesar-rodriguez @massongit

**Comment by Hanse00** [comment]
Minor correction:

> Mega linter has higher repository and container usage stats compared to the ~~megalinter~~ super-linter when looking at recent periods of time

**Comment by nvuillam** [comment]
Thanks for this summary @zkoppert !

Of course if any of these bullet points could be an issue for migrating SL to ML, we can discuss about it :)

**Comment by nvuillam** [comment]
Any opinions here ? :)

Since the latest post, we both released a new versions of MegaLinter and SuperLinter

- https://github.com/oxsecurity/megalinter/releases/tag/v7.4.0
- https://github.com/super-linter/super-linter/releases/tag/v5.3.1

We coud have shared this workload :)

# SuperLinter

🐛 Bug Fixes
- Add exception for type-linting @zkoppert (https://github.com/super-linter/super-linter/pull/4470)
🧰 Maintenance
- Bump react-router-dom from 6.15.0 to 6.16.0 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4670)
- Bump react-intl from 6.4.4 to 6.4.7 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4673)
- Add Babel preset for React package @fdanielsen (https://github.com/super-linter/super-linter/pull/4653)
- Bump configargparse from 1.5.3 to 1.7 in /dependencies/python https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4659)
- Bump docker/build-push-action from 4 to 5 https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4663)
- Bump eslint from 8.48.0 to 8.49.0 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4668)
- Bump stylelint-scss from 5.1.0 to 5.2.1 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4669)
- Bump eslint-plugin-jest from 27.2.3 to 27.4.0 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4671)
- Bump renovate from 36.92.0 to 36.96.8 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4672)
- Bump stylelint-config-standard-scss from 9.0.0 to 11.0.0 in /dependencies https://github.com/dependabot (https://github.com/super-linter/super-linter/pull/4615)
- Remove notice from april availability issue @zkoppert (https://github.com/super-linter/super-linter/pull/4654)
Put dependency updates in the maintenance section by default @zkoppert (https://github.com/super-linter/super-linter/pull/4655)

# MegaLinter

## What's Changed

- Core
  - Upgrade base image to to alpine 3.18 by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2903
  - Upgrade python to 3.11.5 by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2930

- Media
  - [Maximize your code consistency with Megalinter](https://codewithme.cloud/posts/2023/08/maximize-your-code-consistency-with-megalinter/) by [Tor Ivar Asbølmo](https://www.linkedin.com/in/torivara/) on [codewithme.cloud](https://codewithme.cloud)
  - [MegaLinter in Azure DevOps](https://jamescook.dev/megalinter-in-azure-devops) by [James Cook](https://www.linkedin.com/in/cookjames/)

- Fixes
  - Fix [IGNORE_GITIGNORED_FILES not working anymore](https://github.com/oxsecurity/megalinter/issues/2955) , by @iisisrael in https://github.com/oxsecurity/megalinter/pull/2967
  - Fix [v7 issue when using MEGALINTER_FILES_TO_LINT](https://github.com/oxsecurity/megalinter/issues/2744) (thanks @pfiaux !) by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2917
  - Fix [Ignore symlink files when VALIDATE_ALL_CODEBASE is false](https://github.com/oxsecurity/megalinter/issues/2944) by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2964
  - Fix rstcheck options & install by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2899
  - Undowngrade Kubescape  by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2901


- Doc
  - Secretlint logo - reduce size to 150 and remove background by @itayox in https://github.com/oxsecurity/megalinter/pull/2880
  - Replace `https://megalinter.io/flavors` with `https://megalinter.io/latest/flavors` to avoid lychee 404 error by @ashokm in https://github.com/oxsecurity/megalinter/pull/2910
  - docs: config-cli-lint-mode.md: Grammar tweaks by @rasa in https://github.com/oxsecurity/megalinter/pull/2953

- CI
  - Reformatted config files to not fail common linters (closes #2890) by @andrewvaughan in https://github.com/oxsecurity/megalinter/pull/2891
  - Workflow job name changed from `build` to `megalinter` to prevent conflicts with other workflows
  - Add support for master branch in TEMPLATES/mega-linter.yml by @rasa in https://github.com/oxsecurity/megalinter/pull/2888
  - Upgrade editorconfig-checker & setup-node by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2898
 
- Deprecations
  - Deprecate SCSS LINT as not maintained anymore (<https://github.com/sds/scss-lint#notice-consider-other-tools-before-adopting-scss-lint>) by @nvuillam in https://github.com/oxsecurity/megalinter/pull/2902

- Linter versions upgrades
  - [actionlint](https://rhysd.github.io/actionlint/) from 1.6.25 to **1.6.26**
  - [ansible-lint](https://ansible-lint.readthedocs.io/) from 6.17.2 to **6.20.0**
  - [bicep_linter](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/linter) from 0.20.4 to **0.21.1**
  - [black](https://black.readthedocs.io/en/stable/) from 23.7.0 to **23.9.1**
  - [cfn-lint](https://github.com/aws-cloudformation/cfn-lint) from 0.79.6 to **0.80.2**
  - [checkov](https://www.checkov.io/) from 2.3.360 to **2.4.10**
  - [checkstyle](https://checkstyle.sourceforge.io) from 10.12.2 to **10.12.3**
  - [chktex](https://www.nongnu.org/chktex) from 1.7.6 to **1.7.8**
  - [clippy](https://github.com/rust-lang/rust-clippy) from 0.1.71 to **0.1.72**
  - [clj-kondo](https://github.com/borkdude/clj-kondo) from 2023.07.13 to **2023.09.07**
  - [cspell](https://github.com/streetsidesoftware/cspell/tree/master/packages/cspell) from 6.31.3 to **7.3.6**
  - [devskim](https://github.com/microsoft/DevSkim) from 1.0.18 to **1.0.22**
  - [djlint](https://djlint.com/) from 1.32.1 to **1.34.0**
  - [dotnet-format](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-format) from 6.0.413 to **6.0.414**
  - [editorconfig-checker](https://editorconfig-checker.github.io/) from 2.7.0 to **2.7.1**
  - [eslint](https://eslint.org) from 8.46.0 to **8.49.0**
  - [git_diff](https://git-scm.com) from 2.38.5 to **2.40.1**
  - [gitleaks](https://github.com/gitleaks/gitleaks) from 8.17.0 to **8.18.0**
  - [golangci-lint](https://golangci-lint.run/) from 1.54.0 to **1.54.2**
  - [helm](https://helm.sh/docs/helm/helm_lint/) from 3.10.2 to **3.11.3**
  - [jscpd](https://github.com/kucherenko/jscpd/tree/master/packages/jscpd) from 3.5.9 to **3.5.10**
  - [kics](https://www.kics.io) from 1.7.5 to **1.7.8**
  - [ktlint](https://ktlint.github.io) from 0.50.0 to **1.0.0**
  - [kubescape](https://github.com/kubescape/kubescape) from 2.3.6 to **2.9.0**
  - [markdownlint](https://github.com/DavidAnson/markdownlint) from 0.35.0 to **0.36.0**
  - [mypy](https://mypy.readthedocs.io/en/stable/) from 1.4.1 to **1.5.1**
  - [phpstan](https://phpstan.org/) from 1.10.28 to **1.10.35**
  - [powershell](https://github.com/PowerShell/PSScriptAnalyzer) from 7.3.6 to **7.3.7**
  - [powershell_formatter](https://github.com/PowerShell/PSScriptAnalyzer) from 7.3.6 to **7.3.7**
  - [prettier](https://prettier.io/) from 3.0.1 to **3.0.3**
  - [protolint](https://github.com/yoheimuta/protolint) from 0.45.0 to **0.46.0**
  - [psalm](https://psalm.dev) from Psalm.5.14.1@ to **Psalm.5.15.0@**
  - [puppet-lint](http://puppet-lint.com/) from 4.0.1 to **4.2.0**
  - [pyright](https://github.com/Microsoft/pyright) from 1.1.321 to **1.1.327**
  - [revive](https://revive.run/) from 1.3.2 to **1.3.4**
  - [rstcheck](https://github.com/myint/rstcheck) from 6.1.2 to **6.2.0**
  - [rstfmt](https://github.com/dzhu/rstfmt) from 0.0.13 to **0.0.14**
  - [rubocop](https://rubocop.org/) from 1.56.0 to **1.56.3**
  - [ruff](https://github.com/charliermarsh/ruff) from 0.0.284 to **0.0.290**
  - [scalafix](https://scalacenter.github.io/scalafix/) from 0.11.0 to **0.11.1**
  - [semgrep](https://semgrep.dev/) from 1.34.1 to **1.41.0**
  - [sfdx-scanner-apex](https://forcedotcom.github.io/sfdx-scanner/) from 3.15.0 to **3.16.0**
  - [sfdx-scanner-aura](https://forcedotcom.github.io/sfdx-scanner/) from 3.15.0 to **3.16.0**
  - [sfdx-scanner-lwc](https://forcedotcom.github.io/sfdx-scanner/) from 3.15.0 to **3.16.0**
  - [snakemake](https://snakemake.readthedocs.io/en/stable/) from 7.32.3 to **7.32.4**
  - [spectral](https://meta.stoplight.io/docs/spectral/README.md) from 6.10.1 to **6.11.0**
  - [sqlfluff](https://www.sqlfluff.com/) from 2.2.1 to **2.3.2**
  - [stylelint](https://stylelint.io) from 15.10.2 to **15.10.3**
  - [syft](https://github.com/anchore/syft) from 0.86.1 to **0.91.0**
  - [terraform-fmt](https://developer.hashicorp.com/terraform/cli/commands/fmt) from 1.5.4 to **1.5.7**
  - [terragrunt](https://terragrunt.gruntwork.io) from 0.48.6 to **0.51.3**
  - [terrascan](https://runterrascan.io/) from 1.18.2 to **1.18.3**
  - [tflint](https://github.com/terraform-linters/tflint) from 0.47.0 to **0.48.0**
  - [trivy-sbom](https://aquasecurity.github.io/trivy/) from 0.44.0 to **0.45.1**
  - [trivy](https://aquasecurity.github.io/trivy/) from 0.44.0 to **0.45.1**
  - [trufflehog](https://github.com/trufflesecurity/trufflehog) from 3.47.0 to **3.56.1**
  - [vale](https://vale.sh/) from 2.28.1 to **2.29.0**
  - [xmllint](http://xmlsoft.org/xmllint.html) from 21004 to **21104**

**Full Changelog**: https://github.com/oxsecurity/megalinter/compare/v7.3.0...v7.4.0

**Comment by nemchik** [comment]
> * Megalinter is licensed as AGPL3 while the super-linter is MIT

I am not well versed in the differences between the licenses, but I do feel strongly that the source should be open for those who choose to use any linter can choose to review and know what is accessing their code. This seems to be a non issue in relation to SL/ML.

> * Shared workload if we all were only managing 1 project instead of separately

Seems like a win for all involved.

> * Megalinter is in Python and super-linter is in Shell/Bash

There's a few considerations here. I would venture to say that it's more likely to find contributors well versed in python than in bash. To be specific, by "well versed" I mean "knows the quirks and how to work with and around them". I also think that by the nature of python being capable of executing shell commands you should in theory not lose any capabilities (if you really wanted to sledgehammer it you could just have python execute shell commands to run all of the super linter bash scripts, but please don't). That being said, I think the language a project is written and maintained in is entirely up to the maintainers/contributors and their comfort with the chosen language. I have not been active in contributing to super-linter since prior to the org move, so I can't put much weight in the "what should we use going forward" discussion other than my opinions stated in this point, and that personally I like bash and know it very well, and I don't know python, but if everyone else maintaining and contributing wants to use python then I believe it's a good choice for the project.

> * Super linter receives additional features (Advanced security and Actions minutes) from GitHub, while OX-Security sponsors the Megalinter

I would imagine if a merge is to go through and GitHub (company) maintains an interest in an all-in-one linting tool existing and wants to back ML that at least some of the benefits afforded to SL could be offered to ML, but that'll be up to GH staff to decide.

> * Megalinter is maintained by Nicolas as a BDFL model with other support maintainers while super linter is run as a community of equal maintainers

In my view, maintainers and contributors participate for a number of reasons, often including "I wanted project to do/have X so I added it" and sometimes "I like being able to put my name on an accomplishment that helped others" and various other reasons. ML readme notes the history with SL, and is said to have been a fork of SL, but does not have a fork relationship to SL that I can see. That's not an issue, just an observation. SL seemingly started as an internal GH project that garnered attention from contributors, some of which (myself included) were given access to assist in maintaining the repository. Neither org has publicly visible members, so for an end user wondering who actually controls/governs either project, the best bet is to look into the repository itself for documentation such as code owners, contributing guidelines, the readme, and anything that those may link to for additional information on the topic, or make educated guesses based on contributors/commits/merges. Some end users may look at SL as "the official/supported/recommended option" due to it's tied to GH (company). ML could gain that recognition/trust depending on how the merge is handled, or the situation could be viewed as many view Google killing off projects. I'm kind of rambling at this point, just thoughts for others to consider.

> * Super linter is not financially sponsored, and Megalinter sponsorship is tied to Nicolas's github sponsors account

No skin in this, but I would envision ongoing active maintainers to have an interest in either shared finances related to the project, or disbanding any funding in the name of keeping the project as neutral as possible.

> * The transition plan would be a deprecation of super-linter in favor of mega-linter, while porting forward some super linter features that megalinter might be missing

This makes the most sense, but it would be good to spell out exactly what deprecation of SL will mean, specifically for end users. Will SL still be usable at all (and just not updated)? Will users be notified without having to check the repo here? etc.

> * Mega linter has higher repository and container usage stats compared to the super linter when looking at recent periods of time

This definitely shows that end users recognize ML as a viable option, and my above mentioned comment about "the official/supported/recommended option" likely has less weight.

---

In conclusion, I think if the active maintainers of both projects either agree that it's best to merge and continue working, or if the SL team has less interest (or time) for maintaining SL since ML is available, it makes sense to deprecate SL in favor of ML, barring any other conflicts/concerns.

**Comment by echoix** [comment]
> ML readme notes the history with SL, and is said to have been a fork of SL, but does not have a fork relationship to SL that I can see. 

It was more like a hard fork, and going back to the commits of the 4.2.4 version, https://github.com/oxsecurity/megalinter/releases/tag/4.2.4 (commit list https://github.com/oxsecurity/megalinter/commits/9478e8b788a8e2d90d8beb66aba728bd3426fea7), at commit https://github.com/oxsecurity/megalinter/commit/93f66de38a1de8f45f01b42fba03f054ca5ce99a files of super-linter were imported.

It was a project started back during the pandemic.

I couldn't go any further back than that, to see if somewhere else the history was kept or not.

**Comment by andrewvaughan** [comment]
One other consideration on the license - AGPL is a copy left license. Nearly every business with a license policy will prevent copy left licenses from being used in their software stack - I have done so many times in my career. This is why I never use Megalinter in my businesses. People also usually use software that they can learn or use professionally, so switching to a copy left license could be extremely detrimental to super linter. I highly recommend megalinter consider switching to a more permissive license to not alienate businesses (and therefore professionals) that may want to use it. You may even see an uptick in the quality of contributors that join the project.

**Comment by nvuillam** [comment]
@nemchik 

> I am not well versed in the differences between the licenses, but I do feel strongly that the source should be open for those who choose to use any linter can choose to review and know what is accessing their code. This seems to be a non issue in relation to SL/ML.

Basically, it's to avoid an "Amazon style" coup (someone monetizing a SAAS service of MegaLinter)

More info here -> https://megalinter.io/latest/license-explanations/

>  I like bash and know it very well, and I don't know python

I learned python while creating MegaLinter, I can assure you that the language is easier than bash, and we use it mainly for scripting :)

> I would imagine if a merge is to go through and GitHub (company) maintains an interest in an all-in-one linting tool existing and wants to back ML that at least some of the benefits afforded to SL could be offered to ML, but that'll be up to GH staff to decide.

Technically, Microsoft is already indirectly backing MegaLinter, as Microsoft backs OX Security who backs me to insure the maintenance of MegaLinter :) 
- https://www.businessinsider.com/ox-security-used-this-pitch-to-win-the-backing-of-microsofts-m12-2022-9?r=US&IR=T 
- https://www.crunchbase.com/funding_round/ox-security-seed--1e047100

> Megalinter is maintained by Nicolas as a BDFL model with other support maintainers while super linter is run as a community of equal maintainers

Technically I must remain leader as I am contractually bound with OX Security to insure the maintenance. But I have the full trust of OX leaders to handle MegaLinter as I see fit, and my vision is the same from the beginning: Provide the best free and open-source tool possible to the developer community, and that won't change :)
Today, I'm the mostly the only one to validate pull requests, because my great co-maintainers @Kurt-von-Laven @echoix @bdovaz are too shy to merge themselves (despite they know they are totally allowed to do so ^^), so I'd be more than happy to see other maintainers merge Pull Requests after validating between each other, that would save me hours of work :)

> No skin in this, but I would envision ongoing active maintainers to have an interest in either shared finances related to the project, or disbanding any funding in the name of keeping the project as neutral as possible.

As I'm contractually bound with OX, I receive some payment for maintenance, based on the number of stars :) But compared to the hours of work spent on working since the beginning, it would be 1$ by hour of work 🤣 
I would be totally open to discuss with ox about sharing external sponsoring between all maintainers (for the record, we had been nominated to [Microsoft FOSS ](https://github.com/microsoft/foss-fund) and it was decided that if we won the amount would have been equally shared between maintainers. Sadly, we lost 😢 

> This makes the most sense, but it would be good to spell out exactly what deprecation of SL will mean, specifically for end users. Will SL still be usable at all (and just not updated)? Will users be notified without having to check the repo here? etc.

I think we could have a transition phase while both SL & ML would remain usable and maintained, we would ask for key users to switch to ML and if the feedback is successful, we could make official the migration from SL to ML after a long period of deprecation notice, to not mess too much with existing SL users

@andrewvaughan

AGPLV3 prevents redistributing MegaLinter ( or publishing an online service) without showing all the sources.
That does not prevent at all to use MegaLinter in internal CI , and companies to not redistribute their internal pipelines :)

More info here -> https://megalinter.io/latest/license-explanations/

**Comment by cprivitere** [comment]
Seems like these answers preclude a true merger due to OX basically wanting to preclude SAAS competition and leadership changes. Maybe if they can be talked into changing that stance?

**Comment by andrewvaughan** [comment]
> AGPLV3 prevents redistributing MegaLinter ( or publishing an online service) without showing all the sources.
> That does not prevent at all to use MegaLinter in internal CI , and companies to not redistribute their internal pipelines :)
> 
> More info here -> [megalinter.io/latest/license-explanations](https://megalinter.io/latest/license-explanations/)

I get the intent, but most enterprises have broad-stroke OSS license requirements precluding copy-left, regardless of whether it's part of the product of DevOps - I've personally never seen a differentiation in policies allowing for AGPL in non-product integrations (such as deployment pipelines). This is just my experience.

I guess it's a question as to what's more important - preventing commercialization of the product, or creating as broad a user-base as possible. Personally, I always go for the latter, as, if a company like Amazon decides to productionalize an OSS system, they build a dependency and naturally must support (and usually fund) it. The license doesn't allow them to simply create ownership of a fork of it. That said, this is up to each developer to decide... and I do think this is a point where Megalinter and Superlinter seemingly differ significantly.

**Comment by nvuillam** [comment]
@andrewvaughan when I created MegaLinter, I was light years away to imagine that it could be used by big companies and organizations ^^ (not exhaustive list below)

![image](https://github.com/super-linter/super-linter/assets/17500430/b23f7040-cfcd-4ecc-9c06-22d789d97201)

I already had discussions with head of ITs of big companies that were afraid to use MegaLinter internally because of AGPL, and it's to be very clear about that they can that I added a page in ML documentation

There are indeed pros and cons for MIT/Apache vs AGPL... and it already happened that a company I never heard about was selling it SAAS service partially based on MegaLinter... 
Maybe if they asked nicely (or better: by fundind ML) I would have been ok with it, but with all the hours that I spend on MegaLinter, I think it's not fair to redistribute MegaLinter without funding, that's why I prefer to remain in AGPL for now (and it does not seem to bother so many users ^^ )

If SL maintainers decide that copyleft is a no-go, we can of course have a discussion about it :)

**Comment by andrewvaughan** [comment]
> I already had discussions with head of ITs of big companies that were afraid to use MegaLinter internally because of AGPL, and it's to be very clear about that they can that I added a page in ML documentation

Totally get it! Only working from my experience - most companies I have worked for that have had a policy was very, _very_ broad stroked. If the open source was copy-left, you couldn't use it. That's why you might not see some non-tech companies using AGPL or other types of software (this is from my Fortune-100 company days).

Generally the guidance was from legal - "we don't even want to test those waters - if there's an alternative, you can use that, otherwise, we'll pay you to replicate it."

**Comment by Hanse00** [comment]
Licenses are always an interesting question. I can't say I have a clear opinion, but I will share a collection of my (perhaps contradictory? Humans are funny that way) thoughts:

* I agree that it sucks to put in a bunch of work on something, only to find that someone else has managed to turn your work into their business. It's just not very "fair".
* I have no expectation of being rewarded (Monetary or otherwise) for my open source work - It's entirely an altruistic enterprise that happens outside of my paid employment. Meaning that although it definitely would suck, it wouldn't particularly hurt me in any way.
* I believe that it's more useful to focus on the reality in front of us, than all the hypotheticals that could one day happen.
* Many clients I've worked with are fairly risk averse, they'll probably be more likely to adopt a "Copyleft bad" stance (As @andrewvaughan hinted at), rather than paying council to understand the true meaning of any given license.
* Even more clients don't even seem to be aware of the licenses that come attached to the software they use... a concerning reality we don't need to solve for here 😅.

**Comment by assignUser** [comment]
My active contributions where a while ago so I can't say much about the current state of things and will keep to more general thoughts. In principal I think a merge would make sense to unify resources etc. as explained above. :+1: 

I am not a fan of the BDFL model in general and much prefer a more [ASF](https://www.apache.org/theapacheway/) like [structure](https://www.apache.org/foundation/voting.html) that can also be implemented outside of the ASF (see e.g. [substrait](https://substrait.io/governance/)). OX supporting ML in general and @nvuillam directly is great and while there are apparently contractual guarantees to keep ML independent etc. as @nvuillam  mentioned repeatedly I think actually implementing/formalizing some form of OX independent governance would be assuring to current and future contributors. There is a recent example what can happen if things have not been setup in time and now a widely used [R package](https://github.com/Rdatatable/data.table/issues/5656) is kind of in limbo.

License wise I get the concerns mentioned here from both sides but don't have any hard preferences in this case.

**Comment by andrewvaughan** [comment]
> Even more clients don't even seem to be aware of the licenses that come attached to the software they use... a concerning reality we don't need to solve for here 😅.

Something a lot of early-stage CTOs come up against in their first SOC audit. Always a fun scramble!

**Comment by nvuillam** [comment]
> There is a recent example what can happen if things have not been setup in time and now a widely used https://github.com/Rdatatable/data.table/issues/5656 is kind of in limbo.

@assignUser > Can't happen here, there are 3 other maintainers than me that can merge PRs & even release new versions :)
So if I get burnt-out , hit by a bus or abducted by aliens, MegaLinter will survive anyway ;)

**Comment by ferrarimarco** [comment]
Hi folks! After a few internal syncs, we decided not to go for the merge. I'll close this issue to reflect the statement, but I'll not lock the conversation if people want to post here.

**Comment by zkoppert** [comment]
👋 @nvuillam Ultimately we decided to go this way each for our own personal reasons, but collectively we wanted to maintain autonomy to try whatever we like with this project and keep it going in its current form. We are wishing the megalinter well on your journey forward as well as choosing our own path for ourselves. Thanks for reaching out!

**Comment by nvuillam** [comment]
@zkoppert thank you for your response neighbour :)
If you ever change of mind, you know where to find us :)

**Comment by Plaenkler** [comment]
> 👋 @nvuillam Ultimately we decided to go this way each for our own personal reasons, but collectively we wanted to maintain autonomy to try whatever we like with this project and keep it going in its current form. We are wishing the megalinter well on your journey forward as well as choosing our own path for ourselves. Thanks for reaching out!

@zkoppert I personally think the decision is a good one. I am glad that SL continues to exist and would like to thank you all for continuing to work on it.
